### ISSUE TITLE
New API suggestion

### ISSUE THREAD

**Issue by dotansimha** [apollographql/apollo-angular#490]
@Urigo @kamilkisiela @eitanfr @tomermoshe @darkbasic

## Current Implementation

So today most of the apps that using apollo-angular do it in one of the following ways (better to worst)

1. Create a Service that required `Apollo` and exposes a function that use the Apollo instance to fetch the data. Then create a Component and inject the service and call it. The service serves the component only, and does not do the work for other components (this way it's easier to move and reuse the component).
2. Same as one, but share a service between multiple components.
3. Inject `Apollo` directly into the Component, and use it inside the Component's code.

## New API using DI

I would like to suggest a new API, that is similar to option number 1 but uses Angular DI, remove the need for the Service, and helps to create a better best-practice by avoiding option number 2. 

The idea is to take the GraphQL Query and create for it an `InjectionToken`, the use the `InjectionToken` and the query itself to create a Provider.
The Provider is created and injected during the creation of the module, and we use the `InjectionToken` with `@Inject` in the Component.

## Example

This is an example for the usage:

`graphql` is the functions that takes a query and creates `InjectionToken`
`provideGraphql` is the functions that takes a query and `InjectionToken`, and creates the Provider

// users.query.ts
```ts
import gql from 'graphql-tag';
import { graphql, provideGraphql } from 'apollo-angular'; 

const usersQuery = gql(`
    query users {
         users {
               id
               name
          }
    }
`);

export const UsersQuery = graphql(machineListQuery); // this is the token
export const UsersQueryProvider = provideGraphql(UsersQuery, machineListQuery); // provider
```

// my-modules.ts
```ts
@NgModule({
    imports: [ ... ],
    declarations: [ ... ],
    providers: [ UsersQueryProvider ] // import here
})
class MyApp {

}
```

Then, in the Component that needs to use the users query:

// users-list-component.ts
```ts
import { GraphQL } from 'apollo-angular';

interface UsersResponse { // even better if generated
    users: { name: string, id: string }[];
}

@Component({ ... })
class MyComponent implemented NgOnInit {
     constructor(@Inject(UsersQuery) users: GraphQL<UsersResponse>) {}
     
     ngOnInit() {
           this.users.query({}); // you can pass variables here 
     }
}
```

So now the Component injects the GraphQL Query directly and there is no need to create a service. 

## Extensions

Also, I thought that we can implement a mechanism for doing transformations over the response, and decouple it from the implementation of the Component. Something like that:

```ts
export const UsersQueryProvider = provideGraphql(UsersQuery, machineListQuery, {
    transform: obs => {
          return obs.map(response => response.data.users);
    },
});
```

## Implementation

I already created a very basic implementation and tested it on an existing app.

```ts
import { InjectionToken } from '@angular/core';
import { Apollo, QueryRef } from 'apollo-angular';
import { getOperationName } from 'apollo-utilities';
import { FetchResult } from 'apollo-link';
import { Observable } from 'rxjs';

export interface GraphQL<ResultType = any, VariablesType = any> {
  query(variables: VariablesType): QueryRef<ResultType>;
  mutate(variables: VariablesType): Observable<FetchResult<ResultType>>;
}

export function provideGraphQL<ResultType = any, VariablesType = any>(token: InjectionToken<any>, document: any) {
  return {
    provide: token,
    useFactory(apollo: Apollo): GraphQL<ResultType, VariablesType> {
      return {
        query(variables: any = {}): QueryRef<ResultType> {
          return apollo.watchQuery({
            query: document,
            variables,
          });
        },
        mutate(variables: any = {}): Observable<FetchResult<ResultType>> {
          return apollo.mutate({
            mutation: document,
            variables,
          });
        },
      };
    },
    deps: [Apollo],
  };
}

export function graphql(document: any) {
  return new InjectionToken(getOperationName(document));
}

``` 

## Issues so far

When dealing with Angular code and AOT on the same file, and trying to defined the GraphQL query and the token creation on the same file, the AOT compiler [throws this exception](https://github.com/angular/angular/blob/master/aio/content/guide/aot-compiler.md#tagged-template-expressions-are-not-supported), the workaround for this is to use `gql(...)` (as function, not as ES2016 tag).

What do you think?

**Comment by intellix** [comment]
Just wanted to add that we're not using it like you said. There are no services and the components define their own GQL data requirements. I think when you use services then you encourage query re-use and over-fetching until you're back to REST.

Below is how we do 2x components: **MessageList** with **Messages**:

**message.graphql.ts**
```ts
import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';

export interface MessageFragment {
  id: string;
  body: string;
  createdAt: string;
  user: {
    id: string;
    name: string;
  };
}

export const fragment: DocumentNode = gql`
  fragment Message on Message {
    id
    body
    createdAt
    user {
      id
      name
    }
  }
`;
```

**message.component.html:**
```html
<header>{{ message.user.name }} {{ message.createdAt }}</header>
<section>{{ message.body }}</section>
```

**message.component.ts**:
```ts
import { MessageFragment } from './message.graphql.ts';

@Component({
  selector: 'xc-message',
  templateUrl: './message.component.html',
})
export class MessageComponent {
  @Input() message: MessageFragment;
}
```

Then a data fetching component which uses the fragment:

**message-list.graphql.ts:**
```ts
import { fragment, MessageFragment } from '../message/message.graphql';

export interface MessageList {
  messages: {
    pageInfo {
      hasPreviousPage: boolean;
      endCursor: string;
    };
    edges: [{
      node: MessageFragment;
    }];
  };
}

export const query: DocumentNode = gql`
  query MessageList {
    messages {
      pageInfo {
        hasPreviousPage
        endCursor
      }
      edges {
        node {
          ...Message
        }
      }
    }
  }

  ${MessageFragment}
`;
```

**message-list.component.ts:**
```ts
import { query } from './message-list.graphql.ts';

@Component({
  selector: 'xc-message-list',
  templateUrl: './message-list.component.html',
})
export class MessageListComponent implements OnInit {

  private messagesQuery = this.apollo.watchQuery<MessageList>({ query });

  messages$: Observable<any> = this.messagesQuery.valueChanges.pipe(
    map(({ data }) => data.messages.edges.map(edge => edge.node)),
  );

  constructor(private apollo: Apollo) {}
}
```

**message-list.component.html:**
```html
<xc-message [message]="message" *ngFor="let message of messages$ | async"></xc-message>
```

**Comment by darkbasic** [comment]
This is basically use case 3, isn't it?

**Comment by bardiel** [comment]
Basically, I think it is; but also, I think think it is the best way of using Graphql.
Still, I really like what your proposal. IMHO, best case scenario would be to have **both** (or even more) ways of querying; though I would not like it to be the default, nor do I think it should be bundled with **angular-apollo**. I would really like it to be a fully independent and optional additional package.

**Comment by kamilkisiela** [comment]
We released something similar in v1.2.0.

Here's a blogpost about it:
https://medium.com/the-guild/apollo-angular-code-generation-7903da1f8559

And a video to show it in action:
https://www.youtube.com/watch?v=KGBPODrjtKA
